File: Day02/C/main.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

char *load_file(const char *filepath)
{
    // Note: I know the file format is ok, so skipping error checking here
    FILE *f = fopen(filepath, "r");
    fseek(f, 0, SEEK_END);
    long m = ftell(f);
    char *buffer = malloc(m);
    fseek(f, 0, SEEK_SET);
    fread(buffer, 1, m, f);
    fclose(f);
    return buffer;
}

int count_occurence(const char *password, char letter)
{
    size_t count = 0;
    while(*password != '\0') {
        count += *password++ == letter;
    }
    return count;
}

int str_to_int(char *line, int offset, int length)
{
    char temp[length + 1];
    strncpy(temp, line + offset, length);
    temp[length + 1] = '\0';
    return atoi(temp);
}

int index_of(char *line, char delim)
{
    int length = sizeof(line)/sizeof(char);
    for (int i = 0; i < length; i++) {
        if (line[i] == delim)
            return i;
    }
    return -1;
}

char *get_password(char *line, int offset)
{
    int length = sizeof(line)/sizeof(char);
    char password[length - offset];
    strcpy(password, line + offset + 1);
}

int isValidPart1(char *ptr)
{
    char *line = strdup(ptr);

    int idxDash = index_of(line, '-');
    int idxColon = index_of(line, ':');

    int minVal = str_to_int(line, 0, idxColon);
    int maxVal = str_to_int(line, idxDash + 1, idxColon - 1 - idxDash);

    char letter[2];
    strncpy(letter, line + idxColon - 1, 2);
    letter[1] = '\0';

    char *password = get_password(line, idxColon);
    int count = count_occurence(password, *letter);

    free(line);
    
    if (minVal <= count && count <= maxVal)
        return 1;

    return 0;
}

int isValidPart2(char *ptr)
{
    char *line = strdup(ptr);

    int idxDash = index_of(line, '-');
    int idxColon = index_of(line, ':');

    int pos1 = str_to_int(line, 0, idxColon);
    int pos2 = str_to_int(line, idxDash + 1, idxColon - 1 - idxDash);

    char letter[2];
    strncpy(letter, line + idxColon - 1, 2);
    letter[1] = '\0';

    char *password = get_password(line, idxColon);

    char letterPos1[2];
    strncpy(letterPos1, password + pos1, 2);
    letterPos1[1] = '\0';

    char letterPos2[2];
    strncpy(letterPos2, password + pos2, 2);
    letterPos2[1] = '\0';

    int count = 0;

    if (strcmp(letterPos1,letter))
        count++;

    if (strcmp(letterPos2,letter))
        count++;

    free(line);

    if (count == 1)
        return 1;
    return 0;
}

int main(int argc, char **argv)
{
    const char *in_file_path = argv[1];

    char *input = load_file(in_file_path);
    int init_size = strlen(input);
	char delim[] = "\n";

	char *ptr = strtok(input, delim);

    int part1Counter = 0;
    int part2Counter = 0;

	while(ptr != NULL)
	{
		part1Counter += isValidPart1(ptr);
        part2Counter += isValidPart2(ptr);
		ptr = strtok(NULL, delim);
	}

    printf("Part 1 valid passwords: %d\n", part1Counter);
    printf("Part 2 valid passwords: %d\n", part2Counter);

    free(input);
    free(ptr);
}

File: Day02/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func loadFile(filename string) ([]string, error) {
	lines := []string{}
	file, err := os.Open(filename)
	if err != nil {
		return lines, err
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, nil
}

func isValidPart1(line, letter string, min, max int) bool {
	count := strings.Count(line, letter)
	if min <= count && count <= max {
		return true
	}
	return false
}

func isValidPart2(line string, letter byte, pos1, pos2 int) bool {
	count := 0
	if line[pos1] == letter {
		count++
	}
	if line[pos2] == letter {
		count++
	}

	if count == 1 {
		return true
	}
	return false
}

func main() {
	lines, err := loadFile("../input.txt")
	if err != nil {
		fmt.Printf("ERROR: %v\n", err.Error())
		os.Exit(1)
	}

	lineCount := 0
	part1Counter := 0
	part2Counter := 0

	for _, ln := range lines {
		lineCount++
		idxColon := strings.Index(ln, ":")
		leftPart := ln[:idxColon]
		// Not trimming here to leave the space (index 0) at the start, since we want indexing from 1
		rightPart := ln[idxColon+1:]

		idxDash := strings.Index(leftPart, "-")
		idxSpace := strings.Index(leftPart, " ")

		min, _ := strconv.Atoi(leftPart[:idxDash])
		max, _ := strconv.Atoi(leftPart[idxDash+1 : idxSpace])
		letter := leftPart[idxSpace+1]

		if isValidPart1(rightPart, string(letter), min, max) {
			part1Counter++
		}

		if isValidPart2(rightPart, letter, min, max) {
			part2Counter++
		}
	}

	fmt.Printf("Total lines %d. Valid passwords for part 1: %d\n", lineCount, part1Counter)
	fmt.Printf("Total lines %d. Valid passwords for part 2: %d\n", lineCount, part2Counter)
}
